flocker/node/agents/gce.py

```diff
@@ -448,44 +488,58 @@ def create_volume(self, dataset_id, size):
         )
 
     def attach_volume(self, blockdevice_id, attach_to):
-        config = dict(
-            deviceName=blockdevice_id,
-            autoDelete=False,
-            boot=False,
-            source=(
-                "https://www.googleapis.com/compute/v1/projects/%s/zones/%s/"
-                "disks/%s" % (self._project, self._zone, blockdevice_id)
+        with start_action(
+                action_type=u"flocker:node:agents:gce:attach_volume",
+                blockdevice_id=blockdevice_id,
+                attach_to=attach_to,
+        ) as action:
+            config = dict(
+                deviceName=blockdevice_id,
+                autoDelete=False,
+                boot=False,
+                source=(
+                    "https://www.googleapis.com/compute/v1/projects/%s/zones/"
+                    "%s/disks/%s" % (self._project, self._zone, blockdevice_id)
+                )
             )
-        )
-        try:
-            # TODO(mewert): Verify timeout and error conditions.
-            # TODO(mewert): Test what happens when disk is attached RW to a
-            #               different instance, raise the correct error.
-            result = self._do_blocking_operation(
-                self._compute.instances().attachDisk,
-                instance=attach_to,
-                body=config
+            try:
+                # TODO(mewert): Verify timeout and error conditions.
+                # TODO(mewert): Test what happens when disk is attached RW to a
+                #               different instance, raise the correct error.
+                result = self._do_blocking_operation(
+                    self._compute.instances().attachDisk,
+                    instance=attach_to,
+                    body=config
+                )
+            except HttpError as e:
+                if e.resp.status == 400:
+                    # TODO(mewert): verify with the rest API that this is the
+                    # only way to get a 400.
+                    raise UnknownVolume(blockdevice_id)
+                else:
+                    raise e
+            errors = result.get('error', {}).get('errors', [])
+            for e in errors:
+                if e.get('code') == u"RESOURCE_IN_USE_BY_ANOTHER_RESOURCE":
+                    raise AlreadyAttachedVolume(blockdevice_id)
```

Comment:
I got hung up on this for a while:

```
c5133747-8a8f-4659-acb6-2e4701edcb5e
+-- flocker:node:agents:gce:attach_volume@1/started
    |-- attach_to: jenkins-centos7-slave-1
    |-- blockdevice_id: flocker-v1-32da4a17-d6ad-4354-891f-d6b98112ad1e
    `-- timestamp: 2016-03-01 00:54:49.635471
    +-- flocker:node:agents:gce:wait_for_operation@2,1/started
        |-- operation:
            |-- id: 11655645756643125264
            |-- insertTime: 2016-02-29T16:54:49.870-08:00
            |-- kind: compute#operation
            |-- name: operation-1456793689649-52cf23661ef68-233b41fc-763f2729
            |-- operationType: attachDisk
            |-- progress: 0
            |-- selfLink: https://www.googleapis.com/compute/v1/projects/clusterhqjenkins/zones/us-central1-f/operations/operation-1456793689649-52cf23661ef68-233b41fc-763f2729
            |-- startTime: 2016-02-29T16:54:50.216-08:00
            |-- status: PENDING
            |-- targetId: 8348751227309162405
            |-- targetLink: https://www.googleapis.com/compute/v1/projects/clusterhqjenkins/zones/us-central1-f/instances/jenkins-centos7-slave-1
            |-- user: 453036982657-compute@developer.gserviceaccount.com
            `-- zone: https://www.googleapis.com/compute/v1/projects/clusterhqjenkins/zones/us-central1-f
        `-- timestamp: 2016-03-01 00:54:50.420111
        +-- flocker:node:agents:gce:wait_for_operation@2,2/succeeded
            |-- final_operation:
                |-- endTime: 2016-02-29T16:54:50.666-08:00
                |-- error:
                    `-- errors: [{u'message': u"The disk resource 'projects/clusterhqjenkins/zones/us-central1-f/disks/flocker-v1-32da4a17-d6ad-4354-891f-d6b98112ad1e' is already being used by 'projects/clusterhqjenkins/zones/us-central1-f/instances/jenkins-centos7-slave-1'", u'code': u'RESOURCE_IN_USE_BY_ANOTHER_RESOURCE'}]
                |-- httpErrorMessage: BAD REQUEST
                |-- httpErrorStatusCode: 400
                |-- id: 11655645756643125264
                |-- insertTime: 2016-02-29T16:54:49.870-08:00
                |-- kind: compute#operation
                |-- name: operation-1456793689649-52cf23661ef68-233b41fc-763f2729
                |-- operationType: attachDisk
                |-- progress: 100
                |-- selfLink: https://www.googleapis.com/compute/v1/projects/clusterhqjenkins/zones/us-central1-f/operations/operation-1456793689649-52cf23661ef68-233b41fc-763f2729
                |-- startTime: 2016-02-29T16:54:50.216-08:00
                |-- status: DONE
                |-- targetId: 8348751227309162405
                |-- targetLink: https://www.googleapis.com/compute/v1/projects/clusterhqjenkins/zones/us-central1-f/instances/jenkins-centos7-slave-1
                |-- user: 453036982657-compute@developer.gserviceaccount.com
                `-- zone: https://www.googleapis.com/compute/v1/projects/clusterhqjenkins/zones/us-central1-f
            `-- timestamp: 2016-03-01 00:54:51.605244
    +-- flocker:node:agents:gce:attach_volume@3/failed
        |-- exception: flocker.node.agents.blockdevice.AlreadyAttachedVolume
        |-- reason: flocker-v1-32da4a17-d6ad-4354-891f-d6b98112ad1e
        `-- timestamp: 2016-03-01 00:54:51.606140

:

```

Couldn't understand how an HTTP error code was being logged and yet the action succeeded.

Is that how the Python library exposes errors in the operation or is that what I'd expect to see if I interacted with the REST API directly?

And it seems strange that the Python API leaks the HTTP error for the initial API request but not for the operations request.